blog/2023-08-12-why-mlops-sucks/index.md

```diff
@@ -0,0 +1,592 @@
+
+```python
+from transformers import AutoTokenizer
+from optimum.onnxruntime import ORTModelForCausalLM
+
+tokenizer = AutoTokenizer.from_pretrained("./onnx_optimum")
+model = ORTModelForCausalLM.from_pretrained("./onnx_optimum/", use_cache=False, use_io_binding=False)
+inputs = tokenizer("My name is Arthur and I live in", return_tensors="pt")
+gen_tokens = model.generate(**inputs, max_length=16)
+assert model.providers == ['CPUExecutionProvider']
+print(tokenizer.batch_decode(gen_tokens))
+# this takes longer time than expected since running on CPU, but generates properly.
+
+# >>> ['<s> My name is Arthur and I live in a small town in the countr']
+```
+
+But when the provider is `CUDAExecutionProvider` I get random gibberish text generation on inference:
+
+```python
+model = ORTModelForCausalLM.from_pretrained("./onnx_optimum/", use_cache=False, use_io_binding=False, provider="CUDAExecutionProvider")
+# >>> 2023-08-02 19:47:43.534099146 [W:onnxruntime:, session_state.cc:1169 VerifyEachNodeIsAssignedToAnEp] Some nodes were not assigned to the preferred execution providers which may or may not have an negative impact on performance. e.g. ORT explicitly assigns shape related ops to CPU to improve perf.
+2023-08-02 19:47:43.534136078 [W:onnxruntime:, session_state.cc:1171 VerifyEachNodeIsAssignedToAnEp] Rerunning with verbose output on a non-minimal build will show node assignments.
+
+inputs = tokenizer("My name is Arthur and I live in", return_tensors="pt").to("cuda")
+gen_tokens = model.generate(**inputs, max_length=16)
+assert model.providers == ['CUDAExecutionProvider', 'CPUExecutionProvider']
+
+print(tokenizer.batch_decode(gen_tokens))
+# >>> ['<s> My name is Arthur and I live in a<unk><unk><unk><unk><unk><unk>']
+```
+
+even with different temperature and other parameter values it's always giving out random unrelated outputs like the one above when using `CUDAExecutionProvider`
+
+This issue was raised [under optimum repository here](https://github.com/huggingface/optimum/issues/1248).
+
+It looks like the local llama 2 7b model which I converted to huggingface format might've had some bug as it's not happening for other (newly updated) llama hf model from hf,
+
+After quickly running some benchmark benchmark and comparing ONNX format with its raw huggingface variant `daryl149/llama-2-7b-chat-hf` we could see the following results on 128 max lengths set on generation config:
+
+- Prompt: *Why should one run Machine learning model on-premises?*
+    - ONNX inference latency: `2.31s`
+    - Huggingface version latency: `3s`
+
+We can see that the ONNX Llama 2 7b model is ~25% faster compared to huggingface variant 🚀
+
+okay cool but now there's a new issue with tensorrt that’s raised [here](https://github.com/huggingface/optimum/issues/1278)
+
+### Dolly V2 7B by Databricks
+
+Next let’s try with Dolly-v2 7B parameter model released by Databricks. Who’s equivalent optimum-cli command for ONNX conversion would be:
```

Comment:
you had an intro for Llama-2-7B-Chat (linking to https://github.com/facebookresearch/llama & https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/convert_llama_weights_to_hf.py), but missing any info here about how to get Dolly v2 weights :)